invokeai/app/services/shared/sqlite/migrations/migration_1.py

```diff
@@ -0,0 +1,370 @@
+        BEGIN
+            UPDATE session_queue
+            SET updated_at = STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')
+            WHERE item_id = old.item_id;
+        END;
+        """,
+    ]
+
+    for stmt in tables + indices + triggers:
+        cursor.execute(stmt)
+
+
+def _create_workflow_images(cursor: sqlite3.Cursor) -> None:
+    tables = [
+        """--sql
+        CREATE TABLE IF NOT EXISTS workflow_images (
+            workflow_id TEXT NOT NULL,
+            image_name TEXT NOT NULL,
+            created_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+            -- updated via trigger
+            updated_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+            -- Soft delete, currently unused
+            deleted_at DATETIME,
+            -- enforce one-to-many relationship between workflows and images using PK
+            -- (we can extend this to many-to-many later)
+            PRIMARY KEY (image_name),
+            FOREIGN KEY (workflow_id) REFERENCES workflows (workflow_id) ON DELETE CASCADE,
+            FOREIGN KEY (image_name) REFERENCES images (image_name) ON DELETE CASCADE
+        );
+        """
+    ]
+
+    indices = [
+        "CREATE INDEX IF NOT EXISTS idx_workflow_images_workflow_id ON workflow_images (workflow_id);",
+        "CREATE INDEX IF NOT EXISTS idx_workflow_images_workflow_id_created_at ON workflow_images (workflow_id, created_at);",
+    ]
+
+    triggers = [
+        """--sql
+        CREATE TRIGGER IF NOT EXISTS tg_workflow_images_updated_at
+        AFTER UPDATE
+        ON workflow_images FOR EACH ROW
+        BEGIN
+            UPDATE workflow_images SET updated_at = STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')
+                WHERE workflow_id = old.workflow_id AND image_name = old.image_name;
+        END;
+        """
+    ]
+
+    for stmt in tables + indices + triggers:
+        cursor.execute(stmt)
+
+
+def _create_workflows(cursor: sqlite3.Cursor) -> None:
+    tables = [
+        """--sql
+        CREATE TABLE IF NOT EXISTS workflows (
+            workflow TEXT NOT NULL,
+            workflow_id TEXT GENERATED ALWAYS AS (json_extract(workflow, '$.id')) VIRTUAL NOT NULL UNIQUE, -- gets implicit index
+            created_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+            updated_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')) -- updated via trigger
+        );
+        """
+    ]
+
+    triggers = [
+        """--sql
+        CREATE TRIGGER IF NOT EXISTS tg_workflows_updated_at
+        AFTER UPDATE
+        ON workflows FOR EACH ROW
+        BEGIN
+            UPDATE workflows
+            SET updated_at = STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')
+            WHERE workflow_id = old.workflow_id;
+        END;
+        """
+    ]
+
+    for stmt in tables + triggers:
+        cursor.execute(stmt)
+
+
+migration_1 = Migration(db_version=1, app_version="3.4.0", migrate=_migrate)
```

Comment:
What's the reason for storing the app_version? I'm wondering:
1) if it's needed at all, and
2) whether we should instead store the version of the app that _ran_ the migration in the DB

Reply:
Included because @lstein had the app_version in the `model_manager_metadata` table and I made an assumption that there was a need for it.

Also figured it was useful for to know during which app release the db was migrated.

I commented just before your review about a flaw in the design of this: https://github.com/invoke-ai/InvokeAI/pull/5227#issuecomment-1846261023

If we store the app version at all, I think we should store both the "from" and "to" versions. We'd need both of those data to derive meaningful information. For example, if a user installing 3.5.0 from 3.2.0 is different than from 3.4.0, and knowing the version they came from is useful for troubleshooting.

Reply:
I've removed this as I don't think it is useful